### PR TITLE
chore(kafka sink): Add batch option passing to kafka sink

### LINKED ISSUE #3483: Pass batch options to librdkafka
# Current Vector Version
```
0.10.0
```

As kafka does its own batching we don't expose any batching config on the Vector side as that would be redundant. This is a usability challenge and we should allow passing batch configuration options into `librdkafka`.

---

Old issue

### Use-cases
Currently, the Vector docs state that batching in the `kafka` sink is unsupported. However, this would be very useful in order to achieve the highest throughput when dealing with a high amount of data:

1. the larger the batch the higher likelihood of a higher compression ratio
2. amortizes the messaging overhead and eliminates the adverse effect of the round trip time

This is greatly explained here:
https://github.com/edenhill/librdkafka/blob/master/INTRODUCTION.md#performance

For example, Kafka-to-Kafka use-case:
https://vector.dev/guides/integrate/sources/kafka/kafka/
If Vector supported Kafka batching it'd be a really great alternative to Kafka MirrorMaker, Replicator, etc in this Kafka-to-Kafka use-case.

### PR DESCRIPTION
**Pull request by Hoverbear**
Passes batching options on to our rdkafka sink's rdkafka options.

Closes #3483 .

I was able to observe batch lingering with:

```toml
api.enabled = true
api.bind = "0.0.0.0:8686"

[sources.generator]
type = "generator"
lines = ['{ "beep": true }', '{ "beep": true }']
batch_interval = 2
count = 10

[transforms.json_parser]
type = "json_parser"
inputs = ["generator"]

[sinks.kafka]
type = "kafka"
inputs = ["json_parser"]
batch.max_events = 3
batch.timeout_secs = 30
bootstrap_servers = "localhost:9091"
encoding.topic = "text"
topic = "forgotton-potato"
```

### CODE REVIEW

**Review by lukesteensen — APPROVED** [review]
The implementation looks good. I assume we need to update cue data so it gets into the docs?

**Comment by Hoverbear** [comment]
Ah, yes!

File: docs/reference/components/sinks/kafka.cue [hunk 1]
```diff
@@ -16,6 +16,13 @@ components: sinks: kafka: {
 		buffer: enabled:      true
 		healthcheck: enabled: true
 		send: {
+			batch: {
```
Comment (Hoverbear) [inline docs/reference/components/sinks/kafka.cue:19]: @binarylogic When I run this through `check-docs` locally I see:

```cue
components.sinks.kafka.features.send: field `batch` not allowed:
    ./docs/reference.cue:133:12
    ./docs/reference.cue:311:13
    ./docs/reference.cue:311:29
    ./docs/reference.cue:380:16
    ./docs/reference.cue:866:19
    ./docs/reference.cue:869:14
    ./docs/reference/components.cue:4:2
    ./docs/reference/components/sinks/kafka.cue:19:4
```

Which I think is due to https://github.com/timberio/vector/blob/7ff9cbf7bc9d77d3b9f019fee8897065a9434be5/docs/reference.cue#L387-L396

Could you help me understand how to communicate this to cue?

Kafka is optionally a batching stream sink if the user configures it, and has no required `timeout_secs` (defaulting to what rdkafka chooses).

Comment (binarylogic) [reply docs/reference/components/sinks/kafka.cue:19]: So if a user does not configure `batch` settings then it is effectively a streaming sink? No batching at all?

Comment (Hoverbear) [reply docs/reference/components/sinks/kafka.cue:19]: Correct, this does not change the behavior of the sink at all, this just aliases batch options to already configurable kafka options.

Comment (Hoverbear) [reply docs/reference/components/sinks/kafka.cue:19]: @binarylogic Friendly re-ping on this. :)

**Review by binarylogic — APPROVED** [review]
Updated the docs. Looks good.

**Comment by Hoverbear** [comment]
I see https://github.com/timberio/vector/blob/e769bccd78bbfa3098282b85567dad0e4f72763e/src/sinks/file/mod.rs#L538 might have caused the failure in https://github.com/timberio/vector/blob/e769bccd78bbfa3098282b85567dad0e4f72763e/src/sinks/file/mod.rs#L503-L544 observed on commit https://github.com/timberio/vector/pull/5010/commits/9060a78869b758234803892aa969eda23caa0a10.

@lukesteensen This may be a flaky test?

**Comment by binarylogic** [comment]
It appears this caused severe performance degradation. From a user:

> Apparently vector 0.11 resets default buffer/queue options of librdkafka and because of that vector 0.11 sends each event and as a separate Kafka message, while nightly-2020-09-24 sends them in a big batch

Adding the following fixed the issue:

```toml
batch.timeout_secs = 1
batch.max_events = 10000
buffer.type = "memory"
buffer.max_events = 10000
```

**Comment by Hoverbear** [comment]
Ah, hmm, this looks to be because of https://github.com/timberio/vector/blob/deb446ab68b6de6cf3091d61b923481ed10e4237/src/sinks/kafka.rs#L52. We're overriding rdkafka with our normal defaults. I think having that an `Option<BatchConfig>` and would default it to `None` here and help this problem.

**Pull request merged**